waddrmgr/db.go

```diff
@@ -862,14 +1225,108 @@ func existsAddress(tx walletdb.Tx, addressID []byte) bool {
 	return bucket.Get(addrHash[:]) != nil
 }
 
+// fetchAddrAccount returns the account to which the given address belongs to.
+// It looks up the account using the addracctidx index which maps the address
+// hash to it's corresponding account id.
+func fetchAddrAccount(tx walletdb.Tx, addressID []byte) (uint32, error) {
+	bucket := tx.RootBucket().Bucket(addrAcctIdxBucketName)
+
+	addrHash := fastsha256.Sum256(addressID)
+	val := bucket.Get(addrHash[:])
+	if val == nil {
+		str := "account not found"
+		return 0, managerError(ErrAccountNotFound, str, nil)
+	}
+	return binary.LittleEndian.Uint32(val), nil
+}
+
+// fetchAccountAddresses loads information about addresses of an account from the database.
+// The returned value is a slice address rows for each specific address type.
+// The caller should use type assertions to ascertain the types.
+func fetchAccountAddresses(tx walletdb.Tx, account uint32) ([]interface{}, error) {
+	bucket := tx.RootBucket().Bucket(addrBucketName)
+
+	accountBucket := tx.RootBucket().Bucket(addrAcctIdxBucketName).
+		Bucket(uint32ToBytes(account))
+	// if index bucket is missing the account, there hasn't been any address
+	// entries yet
+	if accountBucket == nil {
+		return []interface{}{}, nil
+	}
+
+	numAccountAddrs, err := fetchNumAccountAddrs(tx, account)
+	if err != nil {
+		return []interface{}{}, err
+	}
+
+	var addrs = make([]interface{}, numAccountAddrs)
+	i := 0
+	err = accountBucket.ForEach(func(k, v []byte) error {
+		// Skip buckets.
+		if v == nil {
+			return nil
+		}
+		serializedRow := bucket.Get(k)
+
+		// Deserialize the address row first to determine the field
+		// values.
+		row, err := deserializeAddressRow(k, serializedRow)
+		if err != nil {
+			return err
+		}
+
+		// Filter by account
+		if row.account != account {
+			return nil
+		}
+
+		var addrRow interface{}
+		switch row.addrType {
+		case adtChain:
+			addrRow, err = deserializeChainedAddress(k, row)
+		case adtImport:
+			addrRow, err = deserializeImportedAddress(k, row)
+		case adtScript:
+			addrRow, err = deserializeScriptAddress(k, row)
+		default:
+			str := fmt.Sprintf("unsupported address type '%d'",
+				row.addrType)
+			return managerError(ErrDatabase, str, nil)
+		}
+		if err != nil {
+			return err
+		}
+
+		// Handle index out of range due to invalid metadata
+		if i > len(addrs)-1 {
+			str := "inconsistent address data stored in database"
+			return managerError(ErrDatabase, str, nil)
+		}
+		addrs[i] = addrRow
+		i++
+		return nil
+	})
+	if err != nil {
+		return nil, maybeConvertDbError(err)
+	}
+
+	return addrs, nil
+}
+
 // fetchAllAddresses loads information about all addresses from the database.
 // The returned value is a slice of address rows for each specific address type.
 // The caller should use type assertions to ascertain the types.
 func fetchAllAddresses(tx walletdb.Tx) ([]interface{}, error) {
```

Comment:
Please add a TODO here for an iterator in the same vain as `fetchAllAccounts`.   I know you didn't write this function, but I'd like for it to be called out as something we'll need to address.

It's even more important here than with `fetchAllAccounts` since you could potentially have 2^62-1 - 2^32-2 addresses.